### PR TITLE
generator: check reserved keywords

### PR DESCRIPTION
**Pull request by Shaac**
#38

### CODE REVIEW

File: tools/generator/game.py [hunk 1]
```diff
@@ -349,3 +401,47 @@ def explore(obj, schema, path):
         }
     ],
 }
+
+# Obtained with sorted(set(iorgen.parser_c.KEYWORDS +
+# iorgen.parser_cpp.KEYWORDS + iorgen.parser_csharp.KEYWORDS +
+# iorgen.parser_haskell.KEYWORDS + iorgen.parser_java.KEYWORDS +
+# iorgen.parser_php.KEYWORDS + iorgen.parser_rust.KEYWORDS))
+# Not that python keywords are not present, use keyword.iskeyword
```
Comment (remi-dupre) [inline tools/generator/game.py]:
```suggestion
# Note that python keywords are not present, use keyword.iskeyword
```

**Review by remi-dupre — COMMENTED** [review]
Maybe we could benefit keeping keywords from each language separated? For instance it could help maintenance if a language is updating or if we want to stop supporting a language?

File: tools/generator/game.py [hunk 2]
```diff
@@ -171,6 +173,56 @@ def check_field_unicity(self):
                         .format(origin_type, origin, name, def_type, def_name))
                 used_field_names[name] = (origin_type, origin)
 
+    def check_reserved_keywords(self):
+        '''Check that no identifier is in conflict with a reserved keyword'''
```
Comment (seirl) [inline tools/generator/game.py:177]: Seems like you could easily factor this with a nested function:

```
def check_ident(name, what):
    if conflicts_with_reserved_keyword(name):
        raise GameError("{} '{}': ...".format(what, name))
```

File: tools/generator/game.py [hunk 3]
```diff
@@ -349,3 +401,47 @@ def explore(obj, schema, path):
...
+    'synchronized', 'template', 'then', 'this', 'thread_local', 'throw',
+    'throws', 'trait', 'transient', 'true', 'try', 'type', 'typedef', 'typeid',
+    'typename', 'typeof', 'uint', 'ulong', 'unchecked', 'union', 'unsafe',
+    'unset', 'unsigned', 'unsized', 'use', 'ushort', 'using', 'using static',
+    'var', 'virtual', 'void', 'volatile', 'wchar_t', 'where', 'while', 'xor',
+    'xor_eq', 'yield'
+]
+
+
+def conflicts_with_reserved_keyword(value: str) -> bool:
```
Comment (seirl) [inline tools/generator/game.py]: You could move that directly in the nested function above

**Comment by Shaac** [comment]
> Maybe we could benefit keeping keywords from each language separated? For instance it could help maintenance if a language is updating or if we want to stop supporting a language?

Well it’s already separated lists inside iorgen. To update the list we just need to run the command in comment.

File: tools/generator/game.py (hunk 1 above)
Comment (Shaac) [reply tools/generator/game.py]: fixed

File: tools/generator/game.py (hunk 2 above)
Comment (Shaac) [reply tools/generator/game.py:177]: done

File: tools/generator/game.py (hunk 3 above)
Comment (Shaac) [reply tools/generator/game.py]: done

**Review by seirl — APPROVED** [review]

**Pull request merged**